File: src/com/aopphp/go/injector/PhpDealAssertInjector.java
package com.aopphp.go.injector;

import com.aopphp.go.pattern.CodePattern;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.InjectedLanguagePlaces;
import com.intellij.psi.LanguageInjector;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiLanguageInjectionHost;
import com.intellij.psi.util.PsiTreeUtil;
import com.jetbrains.php.lang.PhpLanguage;
import com.jetbrains.php.lang.documentation.phpdoc.psi.PhpDocComment;
import com.jetbrains.php.lang.documentation.phpdoc.psi.tags.PhpDocParamTag;
import com.jetbrains.php.lang.psi.elements.PhpClass;
import com.jetbrains.php.lang.psi.elements.PhpClassMember;
import com.jetbrains.php.lang.psi.elements.StringLiteralExpression;
import org.jetbrains.annotations.NotNull;

/**
 * PhpDeal injector is responsible to inject PHP syntax into the specific annotations
 */
public class PhpDealAssertInjector implements LanguageInjector {

    private static final String PHP_DEAL_ANNOTATION = "\\PhpDeal\\Annotation";

    @Override
    public void getLanguagesToInject(
            @NotNull PsiLanguageInjectionHost host,
            @NotNull InjectedLanguagePlaces injectionPlacesRegistrar)
    {
        // we accept only PHP literal expressions, such as docBlocks and string for PointcutBuilder->method('<expr>')
        if (!(host instanceof StringLiteralExpression) || !host.isValidHost()) {
            return;
        }
        StringLiteralExpression literalExpression = (StringLiteralExpression) host;
        if (!CodePattern.isInsideAnnotation(literalExpression, PHP_DEAL_ANNOTATION)) {
            return;
        }

        PhpClass classInstance    = null;
        TextRange rangeInsideHost = new TextRange(1, Math.max(literalExpression.getTextLength()-1, 1));
        String prefix = "<?php\n";
        String suffix = ";";

        PhpDocComment phpDocComment = PsiTreeUtil.getParentOfType(literalExpression, PhpDocComment.class);
        if (phpDocComment != null) {
            for (PhpDocParamTag param : phpDocComment.getParamTags()) {
                prefix += "/** @var " + param.getType().toStringResolved() + " $" + param.getVarName() + "*/\n";
            }
            PsiElement owner = phpDocComment.getOwner();
            if (owner instanceof PhpClassMember) {
                classInstance = ((PhpClassMember) owner).getContainingClass();
            } else if (owner instanceof PhpClass) {
                classInstance = (PhpClass) owner;
            }
        }
        if (classInstance != null && !classInstance.isInterface()) {
            prefix += "/** @var " + classInstance.getFQN() + " $this */\n";
            prefix += "/** @var " + classInstance.getFQN() + " $__old */\n";
        }

        prefix += "/** @var mixed $__result */\n";
        prefix += "/** @noinspection PhpVoidFunctionResultUsedInspection */\n";
        prefix += "return ";

        injectionPlacesRegistrar.addPlace(PhpLanguage.INSTANCE, rangeInsideHost, prefix, suffix);
    }
}

File: src/main/kotlin/com/aopphp/go/injector/PhpDealAssertInjector.kt
package com.aopphp.go.injector

import com.aopphp.go.pattern.CodePattern
import com.intellij.lang.injection.MultiHostInjector
import com.intellij.lang.injection.MultiHostRegistrar
import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiLanguageInjectionHost
import com.jetbrains.php.lang.PhpLanguage
import com.jetbrains.php.lang.psi.elements.Method
import com.jetbrains.php.lang.psi.elements.ParameterList
import com.jetbrains.php.lang.psi.elements.PhpAttribute
import com.jetbrains.php.lang.psi.elements.PhpClass
import com.jetbrains.php.lang.psi.elements.PhpClassMember
import com.jetbrains.php.lang.psi.elements.StringLiteralExpression
import kotlin.math.max

/**
 * Injects PHP syntax into string literal arguments of #[\PhpDeal\Annotation\*] attributes,
 * providing $this, $__old, $__result, and parameter type hints in the injected fragment.
 *
 * Uses MultiHostInjector so both single-quoted and double-quoted strings are covered.
 */
class PhpDealAssertInjector : MultiHostInjector {

    override fun getLanguagesToInject(registrar: MultiHostRegistrar, context: PsiElement) {
        if (context !is StringLiteralExpression) return
        if (!CodePattern.isInsidePhpAttribute(context, PHP_DEAL_ANNOTATION_PREFIX)) return

        val range = TextRange(1, max(context.textLength - 1, 1))

        // Navigate up to the attribute's owner (method / function / class)
        val paramList = context.parent as? ParameterList ?: return
        val attribute = paramList.parent as? PhpAttribute ?: return
        val owner = attribute.owner

        var prefix = "<?php\n"

        // Add @var hints for method parameters from the owning method's DocBlock
        if (owner is Method) {
            owner.docComment?.paramTags?.forEach { param ->
                prefix += "/** @var ${param.type.toStringResolved()} \$${param.varName} */\n"
            }
        }

        // Add $this and $__old typed as the containing class
        val classInstance: PhpClass? = when (owner) {
            is PhpClassMember -> owner.containingClass
            is PhpClass -> owner
            else -> null
        }
        if (classInstance != null && !classInstance.isInterface) {
            prefix += "/** @var ${classInstance.fqn} \$this */\n"
            prefix += "/** @var ${classInstance.fqn} \$__old */\n"
        }

        prefix += "/** @var mixed \$__result */\n"
        prefix += "/** @noinspection PhpVoidFunctionResultUsedInspection */\n"
        prefix += "return "

        registrar.startInjecting(PhpLanguage.INSTANCE)
            .addPlace(prefix, ";", context as PsiLanguageInjectionHost, range)
            .doneInjecting()
    }

    override fun elementsToInjectIn(): List<Class<out PsiElement>> =
        listOf(StringLiteralExpression::class.java)

    companion object {
        private const val PHP_DEAL_ANNOTATION_PREFIX = "\\PhpDeal\\Annotation"
    }
}
